verifier/src/lib.rs

```diff
@@ -13,192 +13,126 @@ pub use report_body::{
+pub struct VerificationOutput<T> {
+    // Whether or not this verification was successful.
+    succeeded: Choice,
+    // The value that was used in the verification.
+    value: T,
+}
+
+impl<T> VerificationOutput<T> {
+    /// Create a new instance.
     ///
-    /// The `pad` is the number of spaces to precede each line of the displayed
-    /// representation with.
-    pub fn fmt_padded(&self, f: &mut Formatter, pad: usize) -> core::fmt::Result {
-        let option: Option<T> = Option::<T>::from(self.0.clone());
-        match option {
-            Some(value) => value.fmt_padded(f, pad)?,
-            None => write!(f, "{:pad$}Passed", "")?,
-        }
-        Ok(())
+    /// # Arguments
+    /// * `value` - The value that was used in the verification.
+    /// * `succeeded` - Whether or not the verification succeeded.
+    pub fn new(value: T, succeeded: Choice) -> VerificationOutput<T> {
+        Self { value, succeeded }
     }
+
+    /// Was the verification successful?
+    pub fn is_success(&self) -> Choice {
+        self.succeeded
+    }
+
+    /// Was the verification a failure?
+    pub fn is_failure(&self) -> Choice {
+        !self.succeeded
+    }
+}
+
+/// A helper struct for displaying the verification results.
+#[derive(Debug, Clone)]
+pub struct VerificationTreeDisplay<'a, V, O> {
+    verifier: &'a V,
+    result: VerificationOutput<O>,
 }
 
-impl<'a, T> From<&'a CtOption<T>> for CtOptionDisplay<'a, T> {
-    fn from(ct_option: &'a CtOption<T>) -> Self {
-        Self(ct_option)
+impl<'a, V, O> VerificationTreeDisplay<'a, V, O> {
+    /// Create a new instance.
+    ///
+    /// # Arguments
+    /// * `verifier` - The verifier that was used to perform the verification.
+    /// * `result` - The result of the verification.
+    pub fn new(verifier: &'a V, result: VerificationOutput<O>) -> Self {
+        Self { verifier, result }
     }
 }
 
-impl<'a, T: DisplayableError> Display for CtOptionDisplay<'a, T> {
+impl<'a, V: VerificationMessage<O>, O: Clone> Display for VerificationTreeDisplay<'a, V, O> {
     fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
-        self.fmt_padded(f, 0)
+        self.verifier.fmt_padded(f, 0, &self.result)
+    }
+}
+
+trait VerificationMessage<O> {
+    /// Format this verification phase.
+    ///
+    /// # Arguments
+    /// * `f` - The formatter to use.
+    /// * `pad` - The number of spaces to pad, or indent, the message with.
+    /// * `output` - The output of this verification phase
+    fn fmt_padded(
+        &self,
+        f: &mut Formatter<'_>,
+        pad: usize,
+        output: &VerificationOutput<O>,
+    ) -> core::fmt::Result;
+}
+
+impl<V: Display, O: SpacedStructName + Debug> VerificationMessage<O> for V {
+    fn fmt_padded(
+        &self,
+        f: &mut Formatter<'_>,
+        pad: usize,
+        output: &VerificationOutput<O>,
+    ) -> core::fmt::Result {
+        let is_success = output.is_success();
```

Comment:
This might be locking in an implementation too much. I had 3 copies of this so genericised it.
It may be that we want this to key off of a marker trait